### PR TITLE
Ledger object resolution

### PR DESCRIPTION
**Pull request by domwoe**
Implements #9 as far as it's possible with the current version of indy-node and indy-vdr.
Supersedes #8

### CODE REVIEW

**Review by c2bo — COMMENTED** [review]
I don't have enough experience with rust to make any comments regarding rust specific parts but looks good to me 👍 
The only aspect I am wondering about are the regexes (especially the dangling ?*)

File: indy-didresolver/src/did.rs [hunk 1]
```diff
@@ -33,57 +77,153 @@ impl Schema {
...
         Self {
+            schema_seq_no,
             name,
-            schema_id,
             type_: ObjectCodes::ClaimDef as u8,
         }
     }
+
+    fn from_str(input: &str) -> DidIndyResult<ClaimDef> {
+        let re = Regex::new(r"^([0-9]*)/([a-zA-Z0-9_-]?*)").unwrap();
```
[the remark this answers is not in the document]
Comment (domwoe) [reply indy-didresolver/src/did.rs]: Good points. There were quite some bugs in the patterns ;)
I tried to make it simpler.

File: indy-didresolver/src/did.rs [hunk 2]
```diff
@@ -1,12 +1,56 @@
...
+        }
+    }
+}
 
-//TODO: add other Did requests for SCHEMA ...
 #[derive(Debug, PartialEq)]
-pub enum Type {
-    Nym,
+pub enum ObjectFamily {
+    Anoncreds,
```
[the remark this answers is not in the document]
Comment (domwoe) [reply indy-didresolver/src/did.rs:35]: I feel like "Anoncreds" is mostly used as a single term on its own, but I'm ok with "AnonCreds" as well *shrug* ;)

File: indy-didresolver/src/did_document.rs [hunk 3]
```diff
@@ -1,27 +1,71 @@
...
+use super::error::{DidIndyError, DidIndyResult};
+use super::responses::Endpoint;
+use serde::{Deserialize, Serialize};
+use serde_json::{json, Value};
+
+pub const LEGACY_INDY_SERVICE: &str = "endpoint";
+pub const DID_CORE_CONTEXT: &str = "https://www.w3.org/ns/did/v1";
 
 #[derive(Serialize, PartialEq, Debug)]
 pub struct DidDocumentJson {
```
[the remark this answers is not in the document]
Comment (domwoe) [reply indy-didresolver/src/did_document.rs]: jupp, I thought I removed that...

**Review by c2bo — APPROVED** [review]

File: indy-didresolver/src/did.rs [hunk 4]
```diff
@@ -1,89 +1,188 @@
...
+static NAMESPACE_PATTERN: &str = "((?:[a-z0-9_-]+:?){1,2})";
+// uses base58 alphabet
+static INDY_UNQUALIFIED_DID_PATTERN: &str = "([123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz]{21,22})";
+static OBJECT_FAMILY_PATTERN: &str = "([a-z]*)";
+static OBJECT_FAMILY_VERSION_PATTERN: &str = "([a-zA-Z0-9]*)";
+
+static ANONCREDSV0_OBJECTS_PATTERN: &str =
+    "(SCHEMA|CLAIM_DEF|REV_REG_DEF|REV_REG_ENTRY|REV_REG_DELTA)";
+
+static CLIENT_DEFINED_NAME_PATTERN: &str = "([\\w-]*)";
```
[the remark this answers is not in the document]
Comment (domwoe) [reply indy-didresolver/src/did.rs:19]: This pattern is used for schema, claim def, and rev reg names. These names are called client-defined ... name in the spec.

File: indy-didresolver/src/did.rs [hunk 5]
```diff
@@ -92,73 +191,109 @@ impl RevRegEntry {
...
-                ))),
-                "CLAIM_DEF" => Ok(LedgerObject::ClaimDef(ClaimDef::new(
-                    cap.get(2).ok_or(DidIndyError)?.as_str().to_string(),
-                    cap.get(3).ok_or(DidIndyError)?.as_str().to_string(),
-                ))),
-                "REV_REG_DEF" => unimplemented!("Not yet completed"),
-                "REV_REG_ENTRY" => unimplemented!("Not yet completed"),
-
-                _ => Err(DidIndyError),
+            let object_family_str = cap.get(1).ok_or(DidIndyError::InvalidDidUrl)?.as_str();
```
[the remark this answers is not in the document]
Comment (domwoe) [reply indy-didresolver/src/did.rs:213]: Don't we have an invalid DID Url if a capture group is None ?

File: indy-didresolver/src/did.rs [hunk 6]
```diff
@@ -92,73 +191,109 @@ impl RevRegEntry {
...
-    //TODO: change regex to exclude O,0,I,l
-    let did_regex =
-        Regex::new("did:indy:([a-zA-Z]+|[a-zA-Z]+:[a-zA-Z]+):([a-zA-Z0-9]{21,22})(/.*)?$")
-            .expect("Error in the DID Regex Syntax");
+impl DidUrl {
+    pub fn from_str(input: &str) -> DidIndyResult<DidUrl> {
+        
+        let did_regex = Regex::new(format!(r"{}:{}:{}(.+)?$",DID_INDY_PREFIX, NAMESPACE_PATTERN, INDY_UNQUALIFIED_DID_PATTERN).as_str()).unwrap();
+
+        let input_without_query = input.split("?").collect::<Vec<&str>>()[0];
```
[the remark this answers is not in the document]
Comment (domwoe) [reply indy-didresolver/src/did.rs:270]: Oh, yes.

**Comment by domwoe** [comment]
> I've tested the current PR and the order in the didDoc is mixed up:

Do we care about the ordering? For software it shouldn't matter, but I agree that it looks odd.

For the base diddoc, we could fix it by using a struct (as you did with `DidDocumentJson`) but I'm not sure if the order in the diddoc content would be preserved

File: indy-didresolver/src/did.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (domwoe) [reply indy-didresolver/src/did.rs:213]: The input here is only the path of the DID Url, not the entire DID Url

**Comment by c2bo** [comment]
I tried changing the `serde_json` library to use the feature `preserve_order` to preserve ordering. This breaks indy-vdr though :/ I couldn't get cargo to work with two different featuresets for serde_json without breaking stuff - seems to not be supported currently by cargo?
Is everyone fine with creating a (ordered) struct for the top level DID Document and accept that sub-parts will not be ordered for the time being?

File: indy-didresolver/src/did.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (paulbastian) [reply indy-didresolver/src/did.rs:19]: its a bit unclear who the client is in that sense :P

**Comment by paulbastian** [comment]
shall we make a merge and address it later? I think there is no ad-hoc solution

**Pull request merged**